### ISSUE TITLE
TSConfig ModuleResolution NodeNext Breaks @Apollo/Server Import

### ISSUE THREAD

**Issue by CaseyHaralson** [apollographql/apollo-server#7222]
### Issue Description

When the tsconfig moduleresolution is set to "nodenext", the @apollo/server can't be imported because: The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'.

But, changing the moduleresolution to "node" will make the import work.

Is this an issue with Apollo Server?

### Link to Reproduction

https://github.com/CaseyHaralson/apolloserver-bug

### Reproduction Steps

npm install
npm run build

Then change the moduleresolution type and run the build again.

**Comment by glasser** [comment]
We do literally have a smoke test for this as part of our CI (`smoke-test/nodenext`) but thanks for the reproduction — I'll see what's different between your repro and the smoke test.

**Comment by glasser** [comment]
So, we test with these set:
- `module: "nodeNext"` in tsconfig.json
- `moduleResolution: "nodeNext"` in tsconfig.json
- `type: "module"` in package.json

If I set `module: "nodeNext"` in your tsconfig.json and either set `type: "module"` in your package.json or rename the file to `src/app.mts`, it appears to work properly.

(If I only set `type: "module"` or rename to `app.mts` while leaving it as `module: "CommonJS"`, the `tsc` execution succeeds but creates CJS-style code in `build`, which seems like it's probably not what you want.)

Is `moduleResolution: "nodeNext"` intended to be used when you're not using `module: "nodeNext"`? The TS docs about it are pretty scattered but it's not super clear to me that `module: "CommonJS", moduleResolution: "nodeNext"` is a combination that's supposed to work well. What are you trying to accomplish with this combination?

**Comment by glasser** [comment]
> I started with module CommonJS because that seems to be recommended for node serverside packages

FWIW I saw something about this in at least one place in the TypeScript docs that strikes me as outdated (ie, not updated after TS4.7 with the new nodenext options). (In general the TS docs are excellent in that any given piece of text is well-written, but so much recent functionality is largely only documented in their (excellent) release notes posts...)

Anyway I think we are happy to accept changes to try to improve this combination of options but figuring it out ourselves on the core team is probably not going to be a priority, given how every time we try to improve anything relating to the module system it ends up being a huge wild goose chase...

**Comment by CaseyHaralson** [comment]
Ok, I appreciate you taking a look. Do you want me to close this or keep it open? I'm new to the module system so have no idea how to help at the moment.

**Comment by jfrconley** [comment]
`module: commonjs` and `moduleResolution: nodenext` is a perfectly valid combination. nodenext just follows modern nodejs dependency resolution (supporting the exports field) and CommonJS is still a supported format in nodejs. The issue is that typescript treats .d.ts files with same rules as js files for esm detection. `type: module` means that exported .d.ts files are treated as modules and as such cannot be used with a require statement even if a `main` or `require` export field is provided. The whole ESM situation is an absolute nightmare, but this particular issue is causing some pain for us since we use the exports field, but still compile to commonjs.

**Comment by trevor-scheer** [comment]
@jfrconley thanks for some insight! The whole typescript + cjs/esm situation is rough. I won't be able to address the issue myself in any meaningful timeframe, but I'd welcome a PR that did if you'd like to take that on?

**Comment by benasher44** [comment]
Also check out https://arethetypeswrong.github.io/?p=%40apollo%2Fserver%404.7.1, which shows this issue

**Comment by benasher44** [comment]
@trevor-scheer Yes! @andrewbranch did a great job walking through all the edges in [starting from this issue comment](https://github.com/ProseMirror/prosemirror/issues/1305#issuecomment-1290831333)

**Comment by Cellule** [comment]
I took a stab at fixing the typings straight in the repo here.
See my PR #7614
